src/pykeen/nn/node_piece.py

```diff
@@ -0,0 +1,519 @@
+        )
+        # select anchor IDs with smallest distance
+        return torch.as_tensor(
+            numpy.argpartition(distances, kth=min(k, distances.shape[0]), axis=0)[:k, :].T,
+            dtype=torch.long,
+        )
+
+
+class ScipySparseAnchorSearcher(AnchorSearcher):
+    """Find closest anchors using scipy.sparse."""
+
+    def __init__(self, max_iter: int = 5) -> None:
+        self.max_iter = max_iter
+
+    def __call__(self, edge_index: numpy.ndarray, anchors: numpy.ndarray, k: int) -> numpy.ndarray:  # noqa: D102
+        # infer shape
+        num_entities = edge_index.max().item() + 1
+        # create adjacency matrix
+        adjacency = scipy.sparse.coo_matrix(
+            (
+                numpy.ones_like(edge_index[0], dtype=bool),
+                tuple(edge_index),
+            ),
+            shape=(num_entities, num_entities),
+        )
+        # symmetric + self-loops
+        adjacency = adjacency + adjacency.transpose() + scipy.sparse.eye(num_entities, dtype=bool, format="coo")
+        adjacency = adjacency.tocsr()
+        logger.info(f"Created adjacency matrix: {adjacency}")
+
+        # for each entity, determine anchor pool by BFS
+        num_anchors = len(anchors)
+        pool = numpy.zeros(shape=(num_entities, num_anchors), dtype=bool)
+        reachable = numpy.zeros(shape=(num_entities, num_anchors), dtype=bool)
+        reachable[anchors] = numpy.eye(num_anchors, dtype=bool)
+        final = numpy.zeros(shape=(num_entities,), dtype=bool)
+
+        for _i in range(self.max_iter):
+            # propagate one hop
+            reachable = adjacency.dot(reachable)
+            # copy pool if we have seen enough anchors and have not yet stopped
+            num_reachable = reachable.sum(axis=1)
+            enough = num_reachable >= k
+            mask = enough & ~final
+            pool[mask] = reachable[mask]
+            # stop once we have enough
+            final |= enough
+        del reachable, final
+
+        tokens = numpy.full(shape=(num_entities, k), fill_value=-1)
+        # select from pool
+        # use sparse random sampling due to memory footprint
+        entity_ids, anchor_ids = pool.nonzero()
+        unique_entity_ids, counts = numpy.unique(entity_ids, return_counts=True)
+        assert (counts >= k).all()
+        generator = numpy.random.default_rng()
+        # TODO: is this really doing random selection without replacement?
+        # ~ [0, 1)
+        random_numbers = generator.random(size=(counts.size, k), dtype=numpy.float32)
+        random_numbers = numpy.cumsum(random_numbers, axis=-1)
+        random_numbers /= (random_numbers[:, -1, None] + 1.0e-8)
+        # now random_numbers[:, 0] >= 0, random_numbers[:, -1] = 1.0 - eps
+        intra_offset = numpy.floor(random_numbers * counts[:, None]).astype(int)
+        offset = numpy.cumsum(numpy.r_[0, counts])[:-1, None] + intra_offset
+        tokens[unique_entity_ids] = anchor_ids[offset]
```

Comment:
I am getting an error here at this line when playing with Fb15k237 (500 anchors / 20 per node)
```
IndexError: index 2376023 is out of bounds for axis 0 with size 2376023
```

Reply:
I am still not sure whether this method does the right things :sweat_smile: